Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Hole_filling/Triangulate_hole_polyline.h

```diff
@@ -1194,7 +1203,176 @@ class Triangulate_hole_polyline {
   }
 };
 
-/***********************************************************************************
+#ifndef CGAL_HOLE_FILLING_DO_NOT_USE_CDT2
+/************************************************************************
+ * Triangulate hole by using a cdt_2
+ ************************************************************************/
+// /!\ points.first == points.last
+
+
+template <
+  typename PointRange, //need size()
+  typename Tracer,
+  typename Validity_checker,
+  typename Traits
+>
+bool
+triangulate_hole_polyline_with_cdt(const PointRange& points,
+                          Tracer& tracer,
+                          const Validity_checker& is_valid,
+                          const Traits& traits)
+{
+  typedef typename Traits::FT FT;
+  typedef typename Traits::Point_3 Point_3;
+  typedef typename Traits::Vector_3 Vector_3;
+  typedef typename Traits::Collinear_3 Collinear_3;
+  typedef typename Traits::Compute_squared_length_3 Squared_length_3;
+
+  // Compute an average normal of the hole.
+  const Squared_length_3 squared_length_3 =
+    traits.compute_squared_length_3_object();
+  const Collinear_3 collinear_3 =
+    traits.collinear_3_object();
+
+  std::vector<Point_3> P(boost::begin(points), boost::end(points));
+  if(P.front() != P.back()){
+    P.push_back(P.front());
+  }
+
+  FT x = FT(0), y = FT(0), z = FT(0);
+  std::size_t num_normals = 0;
+  const Point_3& ref_point = P[0];
+  std::size_t size = P.size()-1;
+  for (std::size_t i = 1; i < size; ++i) {
+    const std::size_t ip = i + 1;
+
+    const Point_3& p1 = ref_point; // 3 points, which form a triangle
```

Comment:
Any reason why you use a fixed point for P1? instead of for example taking triples of consecutive points?

Reply:
So, the reason is that if the whole has multiple collinear vertices along its boundary, taking the fixed first point should lead to a higher number of non-degenerate triangles. Moreover, in this case, the generated triangles cover the hole and hence approximate its shape somehow, which should lead to better normal estimation. Using only consecutive points takes into account only the region in the vicinity of the hole boundary, but not the hole-itself.